src/strings/utf8.c

```diff
@@ -384,8 +384,25 @@ MVMuint32 MVM_string_utf8_decodestream(MVMThreadContext *tc, MVMDecodeStream *ds
                     break;
                 }
                 case UTF8_REJECT:
-                    MVM_free(buffer);
-                    MVM_exception_throw_adhoc(tc, "Malformed UTF-8");
+                    if (bufsize >= 3) {
+                        MVMGrapheme32 a = buffer[pos - 2], b = buffer[pos - 1], c = buffer[pos];
+                        MVM_free(buffer);
+                        MVM_exception_throw_adhoc(tc, "Malformed UTF-8 near bytes %02x %02x %02x", a, b, c);
+                    }
+                    else if (bufsize == 2) {
+                        MVMGrapheme32 a = buffer[pos - 1], b = buffer[pos];
+                        MVM_free(buffer);
+                        MVM_exception_throw_adhoc(tc, "Malformed UTF-8 near bytes %02x %02x", a, b);
+                    }
+                    else if (bufsize == 1) {
+                        MVMGrapheme32 a = buffer[pos];
+                        MVM_free(buffer);
+                        MVM_exception_throw_adhoc(tc, "Malformed UTF-8 near byte %02x", a);
+                    }
+                    else {
+                        MVM_free(buffer);
+                        MVM_exception_throw_adhoc(tc, "Malformed UTF-8");
+                    }
```

Comment:
I don't like the repeated free of `buffer` here... you could do a `char first[3]; memcpy(first, buffer, MAX(bufsize, sizeof(first))` and use that? (Alternatively, add `buffer` as `waste` argument to `MVM_exception_throw_adhoc_free`)

Reply:
Addressed in ab86372.